packages/diagram/src/layout/font/subsetFont.ts

```diff
@@ -0,0 +1,329 @@
+            subset = harfbuzzJsWasm.hb_subset_or_fail(face, input);
+            if (subset === 0) {
+                this.cleanupResources(harfbuzzJsWasm, face, fontBuffer);
+                throw new Error(
+                    "hb_subset_or_fail (harfbuzz) returned zero, indicating failure. Maybe the input file is corrupted?"
+                );
+            }
+        } finally {
+            harfbuzzJsWasm.hb_subset_input_destroy(input);
+        }
+
+        const result = harfbuzzJsWasm.hb_face_reference_blob(subset);
+        const offset = harfbuzzJsWasm.hb_blob_get_data(result, 0);
+        const subsetByteLength = harfbuzzJsWasm.hb_blob_get_length(result);
+        if (subsetByteLength === 0) {
+            this.cleanupResources(harfbuzzJsWasm, face, fontBuffer, result, subset);
+            throw new Error("Failed to create subset font, maybe the input file is corrupted?");
+        }
+
+        const subsetFont = Buffer.from(heapu8.subarray(offset, offset + subsetByteLength));
+        this.cleanupResources(harfbuzzJsWasm, face, fontBuffer, result, subset);
+        return subsetFont;
+    }
+
+    /**
+     * Imports a font into harfbuzz
+     *
+     * @param harfbuzzJsWasm the harfbuzzjs instance
+     * @param convertedFont the converted font
+     * @param heapu8 the harfbuzz instance heap to use
+     * @returns the input, face and font buffer
+     */
+    private importFontIntoHarfbuzz(
+        harfbuzzJsWasm: any,
+        convertedFont: Buffer,
+        heapu8: Uint8Array
+    ): { input: Pointer; face: Pointer; fontBuffer: Pointer } {
+        const input: Pointer = harfbuzzJsWasm.hb_subset_input_create_or_fail();
+        if (input === 0) {
+            throw new Error("hb_subset_input_create_or_fail (harfbuzz) returned zero, indicating failure");
+        }
+
+        const fontBuffer: Pointer = harfbuzzJsWasm.malloc(convertedFont.byteLength);
+        heapu8.set(new Uint8Array(convertedFont), fontBuffer);
+
+        const blob: Pointer = harfbuzzJsWasm.hb_blob_create(
+            fontBuffer,
+            convertedFont.byteLength,
+            SubsetManager.HB_MEMORY_MODE_WRITABLE,
+            0,
+            0
+        );
+        const face: Pointer = harfbuzzJsWasm.hb_face_create(blob, 0);
+        harfbuzzJsWasm.hb_blob_destroy(blob);
+        return { input, face, fontBuffer };
+    }
+
+    /**
+     * Sets the input unicodes for the subset
+     *
+     * @param harfbuzzJsWasm the harfbuzzjs instance
+     * @param input the harfbuzz input pointer
+     * @param characters the characters to include in the subset, all characters are included if undefined
+     */
+    private setInputUnicodesForSubset(harfbuzzJsWasm: any, input: Pointer, characters: string | undefined) {
+        const inputUnicodes = harfbuzzJsWasm.hb_subset_input_unicode_set(input);
+        if (characters != undefined) {
+            for (const c of characters) {
+                harfbuzzJsWasm.hb_set_add(inputUnicodes, c.codePointAt(0));
```

Comment:
Erm… This looks like a dangerous assumption that will either break some joke fonts or Unicode support

Reply:
it's based on the library we forked
in particular, we do not set `HB_SUBSET_FLAGS_NO_LAYOUT_CLOSURE`, so ligatures should work
the best implementation would be to collect the required glyphs and use `hb_subset_input_glyph_set`, however this would result in a two-pass subsetting procedure which is imho not worth it
however if you provide me with an example where it does not work, I can change it